### ISSUE TITLE
Caching policy

### ISSUE THREAD

**Issue by dyst5422** [FormidableLabs/urql#40]
I've used Apollo extensively over the last year for almost all of our projects and I think the caching policy they use is very well attuned to the fundamental structure of a node graph.

Each item in the graph is cached according to its typename and id, which means that as long as you have a unique id within a type - something that is a fundamental part of storing data in a database to begin with - then you have a way to tell all nodes apart. GraphQL has built into the specification the best practice of returning mutated items as well, which solves the cache invalidation problem on a very granular level.

Basically, I think the guys over at Apollo nailed caching for GraphQL and it leaves me wondering why urql has decided to do a very expensive greedy invalidation approach. I think there are cases where it might make sense, but in general, the Apollo method is more complete. What might make sense is to have a more configurable caching policy. The current method of hashing just the query and variables prevents more advanced caching policies like this.

Thoughts?

**Comment by kenwheeler** [comment]
On what grounds would you consider the approach expensive? Especially more expensive than storing each item. Greedy invalidation by default, sure, but with a nice escape hatch for granular invalidation.

**Comment by kenwheeler** [comment]
Did you happen to see the shouldInvalidate option?

**Comment by kenwheeler** [comment]
shouldInvalidate let's you invalidate on granular criteria. The cache render prop let's you run an update against your existing cache. Mine is a bit more explicit in that sense, but skips the whole normalization/denormalization step, and gives predictability with regards to nesting

**Comment by dyst5422** [comment]
So when I know part of a query should be invalidated, say one of my 1000 nodes. Then I would return true in shouldInvalidate. Then wouldn't it pull all 1000 nodes from the server again? How would I update just that one item returned in the mutation without doing a full 1000 nodes refetch?

Not trying to criticize. This just happens to be by far the most common operation I perform - update one node on mutation return belonging to large sets in potentially multiple queries (where the full fetch can take multiple seconds to complete). I think it's a compelling use case for what I do.

**Comment by kenwheeler** [comment]
Using the `cache` render prop's `update` method
